### PR TITLE
Staking module: approximate fraction into perbill to avoid overflow

### PR DESCRIPTION
**Pull request by thiolliere**
previous strategy: https://github.com/paritytech/substrate/pull/2863
related to #706 and #1572

in this PR we solve the potential overflow by approximating the fraction (nominator.value/total_exposure) into a Perbill.

This approximation is visible in tests as you can see a -1 in some reward in some test.

TODO:
* [x] maybe implement from rational_approximation for other Perthings
* [x] test for perthing
* do we accept such approximation everytime or should we approximate only when it overflows ?
* maybe use a more precise approximation like PerU64 (it is the maximum we can do as otherwise we don't have double size arithmetic in our hands) ?

### CODE REVIEW

**Comment by Kianenigma** [comment]
This is good in terms of replacing the two potentially dangerous multiplication closures. I wonder if we have anywhere else in runtime such cases? (relating to the topic of the issue: _sweep runtime.._).

Furthermore: 

I think this is enough to replace the whole `ACCURACY` constant in phragmen. I can do it myself by pushing here or you can take over. 

```diff
if *c != n.who {
-	let ratio = {
-		// Full support. No need to calculate.
-		if *n.load == *e.load { Perbill::one() }
-		else {
-			// This should not saturate. Safest is to just check
-			if let Some(r) = ACCURACY.checked_mul(*e.load) {
-				r / n.load.max(1)
-			} else {
-				// Just a simple trick.
-				*e.load / (n.load.max(1) / ACCURACY)
-			}
-		}
-	};
+      let ratio = Perbill::from_rational_approximation(*e.load() / *n.load)
	e.ratio = ratio;
	assignment.1.push((e.who.clone(), ratio));
}
```

Since here the ACCURACY is already u32 it should be good to fit in Perbill. Some minor changes are then needed to the lib.rs file as well.

I also thought about what to do with U128. I think unfortunately it is simply impossible to approximate it with Perbill/mill due to notorious 

```
// tries to approximate `1/approval_stake`
c.score = Fraction::from_xth(c.approval_stake);
``` 

where c.approval_stake can, in theory (not in practice thou') be more than the max balance type and this is will be in an issue: 

```
Perbill::from_rational_approximation(1, 3*u64::max_value()), 
```

(^^ actually overflows, I was expecting it to collapse to zero)

File: core/sr-primitives/src/lib.rs [hunk 1]
```diff
@@ -658,6 +687,20 @@ mod tests {
 			assert_eq!($per::from_percent(1) * $num_type::max_value(), $num_type::max_value() / 100);
 			assert_eq!($per::from_percent(0) * $num_type::max_value(), 0);
 
+			// from_rational_approximation
```
Comment (Kianenigma) [inline core/sr-primitives/src/lib.rs]: These tests + their name in macro suggest testing _multiplication_. Maybe separate them or rename it to sth more general?

**Comment by Kianenigma** [comment]
To further demonstrate the issue with why I needed `PerU128` and its `from_xth()` back then: 

```
assert_eq!(
			Perbill::from_rational_approximation(1_u32, 1_000_000_000_u32), 
			Perbill::from_parts(1),
		);

		assert_eq!(
			Perbill::from_rational_approximation(2_u32, 2_000_000_001_u32), 
			Perbill::from_parts(1),
		);
```

Though I wonder if we can come up with ways to do the second test properly.

**Comment by thiolliere** [comment]
Yes I think introducing Persomething in phragmen worth it.

we can introduce Perbill or create PerU32 or even create PerU64 (we can't do more easily as we need double size arithmetic, and u128 is kind of the limit.

**Comment by thiolliere** [comment]
though in another PR

**Comment by gavofyork** [comment]
would be nice to see a test for the overflow case (that fails on the original code)

**Comment by thiolliere** [comment]
yes done

File: core/sr-primitives/src/lib.rs [hunk 2]
```diff
@@ -162,6 +162,21 @@ impl Permill {
 	/// Converts a fraction into `Permill`.
 	#[cfg(feature = "std")]
 	pub fn from_fraction(x: f64) -> Self { Self((x * 1_000_000.0) as u32) }
+
+	/// Approximate the fracion `p/q` into a per million fraction
```
Comment (niklasad1) [inline core/sr-primitives/src/lib.rs]:
```suggestion
	/// Approximate the fraction `p/q` into a per million fraction
```

File: core/sr-primitives/src/lib.rs [hunk 3]
```diff
@@ -254,6 +269,21 @@ impl Perbill {
 	#[cfg(feature = "std")]
 	/// Construct new instance whose value is equal to `x` (between 0 and 1).
 	pub fn from_fraction(x: f64) -> Self { Self((x.max(0.0).min(1.0) * 1_000_000_000.0) as u32) }
+
+	/// Approximate the fracion `p/q` into a per billion fraction
```
Comment (niklasad1) [inline core/sr-primitives/src/lib.rs]:
```suggestion
	/// Approximate the fraction `p/q` into a per billion fraction
```

**Pull request merged**